From 8e38a920d2601cca320dc6225e084528ff9e415c Mon Sep 17 00:00:00 2001
From: Stefan Hajnoczi <stefanha@redhat.com>
Date: Tue, 13 Aug 2013 09:06:34 +0200
Subject: vhost: clear signalled_used_valid on vhost stop

RH-Author: Stefan Hajnoczi <stefanha@redhat.com>
Message-id: <1376384797-4701-4-git-send-email-stefanha@redhat.com>
Patchwork-id: 53208
O-Subject: [PATCH v2 3/6] vhost: clear signalled_used_valid on vhost stop
Bugzilla: 995030
RH-Acked-by: Laszlo Ersek <lersek@redhat.com>
RH-Acked-by: Kevin Wolf <kwolf@redhat.com>
RH-Acked-by: Michael S. Tsirkin <mst@redhat.com>

From: "Michael S. Tsirkin" <mst@redhat.com>

When vhost device stops, its implementation synchronizes kernel state
back to virtio.c so we can continue emulating the device
in userspace.

This patch ensures that virtio.c's signalled_used_valid flag is reset so
that userspace does not suppress guest notifications due to stale
signalled_used values.

Cc: qemu-stable@nongnu.org
Signed-off-by: Michael S. Tsirkin <mst@redhat.com>
(cherry picked from commit 3561ba14188b3c1e54246ed6db97896bbc082d2f)
Signed-off-by: Stefan Hajnoczi <stefanha@redhat.com>

diff --git a/hw/virtio/vhost.c b/hw/virtio/vhost.c
index fbabf99..0dabf26 100644
--- a/hw/virtio/vhost.c
+++ b/hw/virtio/vhost.c
@@ -761,6 +761,7 @@ static void vhost_virtqueue_stop(struct vhost_dev *dev,
         fflush(stderr);
     }
     virtio_queue_set_last_avail_idx(vdev, idx, state.num);
+    virtio_queue_invalidate_signalled_used(vdev, idx);
     assert (r >= 0);
     cpu_physical_memory_unmap(vq->ring, virtio_queue_get_ring_size(vdev, idx),
                               0, virtio_queue_get_ring_size(vdev, idx));
